### PR TITLE
Refactored duplicated code and method for Privatekey

### PR DESCRIPTION
**Pull request by ChrisRo89**
@olamy Thanks for your feedback, I changed the things. Of course a star in the import statement is not best practice.

### CODE REVIEW

File: src/main/java/winstone/AbstractSecuredConnectorFactory.java [hunk 1]
```diff
@@ -7,6 +7,14 @@
 
 package winstone;
 
+import org.bouncycastle.*;
```
Comment (olamy) [inline src/main/java/winstone/AbstractSecuredConnectorFactory.java]: no stars import please

File: src/main/java/winstone/AbstractSecuredConnectorFactory.java [hunk 2]
```diff
@@ -122,44 +129,30 @@ protected void configureSsl( Map args, Server server ) throws IOException
...
-            Logger.log( Level.WARNING, SSL_RESOURCES, "HttpsConnectorFactory.LoadPrivateKey" );
+        PEMParser pemParser = new PEMParser(reader);
+        PKCS8EncodedKeySpec privateKey = null;
+
+        try {
+            PEMKeyPair pemKeyPair = (PEMKeyPair) pemParser.readObject();
+            privateKey = new PKCS8EncodedKeySpec(pemKeyPair.getPrivateKeyInfo().getEncoded());
+        }
+        catch (IOException ioex) {
+            new IOException("Failed to read PrivateKey").initCause(ioex);
```
Comment (olamy) [inline src/main/java/winstone/AbstractSecuredConnectorFactory.java]: so we ignore exception?

File: src/main/java/winstone/AbstractSecuredConnectorFactory.java [hunk 3]
```diff
@@ -122,44 +129,30 @@ protected void configureSsl( Map args, Server server ) throws IOException
...
+        }
 
-            KeyFactory kf = KeyFactory.getInstance( "RSA" );
-            return kf.generatePrivate( new RSAPrivateKeySpec( mod, privExpo ) );
+        PrivateKey privateKeyOutput = null;
+        try {
+            privateKeyOutput = KeyFactory.getInstance("RSA").generatePrivate(privateKey);
+        } catch (InvalidKeySpecException e) {
+            e.printStackTrace();
```
Comment (olamy) [inline src/main/java/winstone/AbstractSecuredConnectorFactory.java]: so we ignore exception?

File: src/main/java/winstone/AbstractSecuredConnectorFactory.java [hunk 4]
```diff
@@ -122,44 +129,30 @@ protected void configureSsl( Map args, Server server ) throws IOException
...
+        } catch (NoSuchAlgorithmException e) {
+            e.printStackTrace();
```
Comment (olamy) [inline src/main/java/winstone/AbstractSecuredConnectorFactory.java]: so we ignore exception?

**Comment by ChrisRo89** [comment]
@olamy Everything fine now? :-)

File: src/main/java/winstone/AbstractSecuredConnectorFactory.java [hunk 5]
```diff
@@ -7,30 +7,25 @@
...
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.Reader;
-import java.lang.reflect.Method;
-import java.math.BigInteger;
-import java.security.GeneralSecurityException;
-import java.security.KeyFactory;
-import java.security.KeyStore;
-import java.security.PrivateKey;
+import java.security.*;
```
Comment (olamy) [inline src/main/java/winstone/AbstractSecuredConnectorFactory.java]: no star import please

File: src/main/java/winstone/cmdline/Option.java [hunk 6]
```diff
@@ -58,6 +58,7 @@
     public static final OInt HTTP_PORT=integer("http"+_PORT,8080);
     public static final OString HTTP_LISTEN_ADDRESS=string("http"+ _LISTEN_ADDRESS);
     public static final OInt HTTP_KEEP_ALIVE_TIMEOUT=integer("http" + _KEEP_ALIVE_TIMEOUT, _KEEP_ALIVE_TIMEOUT.defaultValue);
+    public static final OBoolean HTTP_HTTPS_REDIRECT=bool("https_redirect", false);
```
Comment (olamy) [inline src/main/java/winstone/cmdline/Option.java]: is it used?

File: src/main/java/winstone/ServerConnectorFactory.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2003-2006 Rick Knowles <winstone-devel at lists sourceforge net>
+ * Distributed under the terms of either:
+ * - the common development and distribution license (CDDL), v1.0; or
+ * - the GNU Lesser General Public License, v2.1 or later
+ */
+package winstone;
+
+import org.eclipse.jetty.server.*;
```
Comment (olamy) [inline src/main/java/winstone/ServerConnectorFactory.java]: no star import

File: src/main/java/winstone/AbstractSecuredConnectorFactory.java [hunk 8]
```diff
@@ -122,44 +117,30 @@ protected void configureSsl( Map args, Server server ) throws IOException
...
-            Logger.log( Level.WARNING, SSL_RESOURCES, "HttpsConnectorFactory.LoadPrivateKey" );
+        PEMParser pemParser = new PEMParser(reader);
+        PKCS8EncodedKeySpec privateKey = null;
+
+        try {
+            PEMKeyPair pemKeyPair = (PEMKeyPair) pemParser.readObject();
+            privateKey = new PKCS8EncodedKeySpec(pemKeyPair.getPrivateKeyInfo().getEncoded());
+        }
+        catch (IOException ioex) {
+            Logger.log(Level.WARNING, SSL_RESOURCES, ioex.getMessage());
```
Comment (olamy) [inline src/main/java/winstone/AbstractSecuredConnectorFactory.java]: do we really want to ignore such errors? maybe `throw new WinstoneException`

File: src/main/java/winstone/AbstractSecuredConnectorFactory.java [hunk 9]
```diff
@@ -122,44 +117,30 @@ protected void configureSsl( Map args, Server server ) throws IOException
...
+        }
 
-            KeyFactory kf = KeyFactory.getInstance( "RSA" );
-            return kf.generatePrivate( new RSAPrivateKeySpec( mod, privExpo ) );
+        PrivateKey privateKeyOutput = null;
+        try {
+            privateKeyOutput = KeyFactory.getInstance("RSA").generatePrivate(privateKey);
+        } catch (InvalidKeySpecException e) {
+            Logger.log(Level.WARNING, SSL_RESOURCES, e.getMessage());
```
Comment (olamy) [inline src/main/java/winstone/AbstractSecuredConnectorFactory.java]: please factorise this.
do we really want to ignore such errors? maybe `throw new WinstoneException`

File: src/main/java/winstone/AbstractSecuredConnectorFactory.java [hunk 10]
```diff
@@ -122,44 +117,30 @@ protected void configureSsl( Map args, Server server ) throws IOException
...
+        } catch (NoSuchAlgorithmException e) {
+            Logger.log(Level.WARNING, SSL_RESOURCES, e.getMessage());
         }
+
+        return  privateKeyOutput;
```
Comment (olamy) [inline src/main/java/winstone/AbstractSecuredConnectorFactory.java]: as we ignore errors. we will return null?

File: src/main/java/winstone/cmdline/Option.java (hunk 6 above)
Comment (ChrisRo89) [reply src/main/java/winstone/cmdline/Option.java]: With all my changes I forgot to add the use for it :-( now it is added

Comment (ChrisRo89) [reply src/main/java/winstone/cmdline/Option.java]: Thank you :-)

File: src/main/java/winstone/AbstractSecuredConnectorFactory.java (hunk 8 above)
Comment (ChrisRo89) [reply src/main/java/winstone/AbstractSecuredConnectorFactory.java]: No, you're right - I'll throw this exception. Thank you

File: src/main/java/winstone/entities/User.java [hunk 11]
```diff
@@ -0,0 +1,47 @@
+package winstone.entities;
+
+import javax.xml.bind.annotation.*;
```
Comment (olamy) [inline src/main/java/winstone/entities/User.java]: no star import

File: src/main/java/winstone/HttpsConnectorFactory.java [hunk 12]
```diff
@@ -6,15 +6,16 @@
  */
 package winstone;
 
-import org.eclipse.jetty.server.ForwardedRequestCustomizer;
-import org.eclipse.jetty.server.HttpConfiguration;
-import org.eclipse.jetty.server.HttpConnectionFactory;
+import org.eclipse.jetty.server.Handler;
 import org.eclipse.jetty.server.Server;
-import org.eclipse.jetty.server.ServerConnector;
+import org.eclipse.jetty.server.handler.*;
```
Comment (olamy) [inline src/main/java/winstone/HttpsConnectorFactory.java]: no star import

File: src/main/java/winstone/ServerConnectorFactory.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+import org.eclipse.jetty.server.HttpConfiguration;
+import org.eclipse.jetty.server.HttpConnectionFactory;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.ServerConnector;
+import org.eclipse.jetty.util.ssl.SslContextFactory;
+import winstone.cmdline.Option;
+
+import java.util.Map;
+
+public class ServerConnectorFactory {
```
Comment (olamy) [inline src/main/java/winstone/ServerConnectorFactory.java]: I would prefer using Builder pattern and not holding the reference to the ServerConnector

Comment (ChrisRo89) [reply src/main/java/winstone/ServerConnectorFactory.java]: Thank you for the hint - that's better for this case.

File: .idea/compiler.xml [hunk 14]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
```
Comment (jglick) [inline .idea/compiler.xml]: Not sure what all this stuff is but please delete & ignore all of `.idea/`.

Comment (ChrisRo89) [reply .idea/compiler.xml]: Yes, it is for general code style and cleanup and extending for http to https redirect

**Comment by ChrisRo89** [comment]
@olamy I would highly appreciated when you have time for a review, I would not do this PR, when I would not think it could improve Jenkins :-)

File: src/main/java/winstone/HttpsConnectorFactory.java [hunk 15]
```diff
@@ -24,39 +21,38 @@
...
  */
-public class HttpsConnectorFactory extends AbstractSecuredConnectorFactory implements ConnectorFactory {
+public class HttpsConnectorFactory extends AbstractSecuredConnectorFactory {
 
     public boolean start(Map args, Server server) throws IOException {
+
         int listenPort = Option.HTTPS_PORT.get(args);
         String listenAddress = Option.HTTPS_LISTEN_ADDRESS.get(args);
         int keepAliveTimeout = Option.HTTPS_KEEP_ALIVE_TIMEOUT.get(args);
+        boolean redirectToHttps = Option.HTTP_HTTPS_REDIRECT.get(args);
```
Comment (olamy) [inline src/main/java/winstone/HttpsConnectorFactory.java]: you are adding a feature but it's not documented.

File: src/main/java/winstone/cmdline/Option.java (hunk 6 above)
Comment (olamy) [inline src/main/java/winstone/cmdline/Option.java]: this new option must be documented in README and https://github.com/jenkinsci/winstone/blob/master/src/main/resources/winstone/LocalStrings.properties
And part of your commit message

File: src/main/java/winstone/tools/WinstoneControl.java [hunk 16]
```diff
@@ -107,6 +114,6 @@ else if (operation.toLowerCase().startsWith(OPERATION_RELOAD.toLowerCase())) {
      * Displays the usage message
      */
     private static void printUsage() {
-        System.out.println(TOOLS_RESOURCES.getString("WinstoneControl.Usage"));
+        Logger.log(Logger.INFO, TOOLS_RESOURCES, "WinstoneControl.Usage");
```
Comment (olamy) [inline src/main/java/winstone/tools/WinstoneControl.java]: I would prefer keep using System.out if there was any issue configuring Looger....

File: src/main/java/winstone/entities/User.java [hunk 17]
```diff
@@ -0,0 +1,49 @@
+package winstone.entities;
+
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlRootElement;
+
+
+@XmlRootElement(name = "user")
```
[the remark this answers is not in the document]
Comment (ChrisRo89) [reply src/main/java/winstone/entities/User.java]: It's a start to replace the XML processing in the FileRealm class, but I didn't found a hint how the XML structure looks on the end. Is there some template, or XSD? Do you agree for replacing this?

File: pom.xml [hunk 18]
```diff
@@ -290,13 +290,26 @@
...
-      <groupId>org.eclipse.jetty</groupId>
-      <artifactId>jetty-ajp</artifactId>
-      <version>${jetty.version}</version>
+      <groupId>org.bouncycastle</groupId>
+      <artifactId>bcpkix-jdk15on</artifactId>
+      <version>1.62</version>
     </dependency>
-    -->
+      <dependency>
+          <groupId>org.jetbrains</groupId>
```
[the remark this answers is not in the document]
Comment (ChrisRo89) [reply pom.xml]: I removed this one, probably I missed some conifg in my IntelliJ IDEA

File: src/main/java/winstone/accesslog/SimpleAccessLogger.java [hunk 19]
```diff
@@ -142,7 +143,9 @@ protected void doStop() throws Exception {
         if (this.outStream != null) {
             try {
                 this.outStream.close();
-            } catch (IOException err) {}
+            } catch (IOException err) {
+                throw new WinstoneException(err.getMessage(), err);
```
[the remark this answers is not in the document]
Comment (ChrisRo89) [reply src/main/java/winstone/accesslog/SimpleAccessLogger.java]: Changed it to logging that there is at least a error logging - you are right it makes no sense to throw some exception here

**Comment by ChrisRo89** [comment]
> Pretty much unreviewable. Try starting with small, targeted PRs.

**Pull request closed without merging**